File: go-version/leetcode/LC440/find_kth_number.go
package LC440

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/6/9 下午11:37
 * @FilePath: find_kth_number
 * @Description:
 */

func findKthNumber(n, k int) int {
	cur := 1
	k--
	for k > 0 {
		step := getSteps(cur, int64(n))
		if step <= k {
			k -= step
			cur++
		} else {
			k--
			cur *= 10
		}
	}
	return cur
}

func getSteps(cur int, n int64) int {
	step := 0
	first, last := int64(cur), int64(cur)
	for first <= n {
		step += int((min(last, n) - first) + 1)
		first *= 10
		last = last*10 + 9
	}
	return step
}

File: python-version/LC440/find_kth_number.py
class Solution:
    def findKthNumber(self, n: int, k: int) -> int:
        cur = 1
        k -= 1
        while k > 0:
            step = self.get_steps(cur, n)
            if k >= step:
                k -= step
                cur += 1
            else:
                k -= 1
                cur *= 10
        return cur

    @staticmethod
    def get_steps(cur, n):
        step = 0
        first, last = cur, cur
        while first <= n:
            step += (min(last, n) - first) + 1
            first *= 10
            last = last * 10 + 9
        return step
